racket/collects/racket/match/compiler.rkt

```diff
@@ -551,5 +562,167 @@
+  ;;        rhs)] ...
+  ;;     [else (next)]))
+  ;;
+  ;; The syntax-parameterization is done to support (failure-cont).
+  ;; In match, failure-cont would yield to the next clause.
+  ;; In our scheme, however, all case-pattern clauses will be disjoint,
+  ;; so we can skip to the else clause (next) right away.
+  ;;
+  ;; Disjointness across clauses is therefore very important:
+  ;;
+  ;; (match 1 [1 (failure-cont)] [1 (displayln 'hello)])
+  ;;
+  ;; should print "hello". A compilation that doesn't take disjointness into
+  ;; account, such as:
+  ;;
+  ;; (let ([next (λ () (match id))])
+  ;;   (case id
+  ;;     [(1)
+  ;;      (syntax-parameterize ([fail (make-rename-transformer #'next)])
+  ;;        (fail))]
+  ;;     [(1)
+  ;;      (syntax-parameterize ([fail (make-rename-transformer #'next)])
+  ;;        (displayln 'hello))]
+  ;;     [else (next)]))
+  ;;
+  ;; would be incorrect.
+  ;;
+  ;; (=> exit-id) is handled regularly, but with the target being the else clause
+  ;; due to the same reason as above.
+  ;;
+  (cond
+    ;; |vars| = 1
+    [(and (pair? vars) (null? (cdr vars)))
+     (define var (car vars))
+
+     ;; Exact values are those parsed by parse-literal.
+     ;; They all can be handled by `case`.
+     ;;
+     ;; Note that `case` can handle stuff like list, vector, and hash literals too,
+     ;; but match parsing doesn't retain enough information to make
+     ;; the compilation easy, so we'll skip them (at least for now).
+     (let loop ([rows rows] [case-rows '()] [match-rows '()])
+       (cond
+         [(null? rows)
+          (generate-case+match var esc (reverse case-rows) (reverse match-rows))]
+         [else
+          (define row (car rows))
+          (define next-rows (cdr rows))
+          (cond
+            ;; |(Row-pats row)| = |vars| = 1, so we can call Row-first-pat.
+            [(flatten-Or-Exact (Row-first-pat row))
+             =>
+             (λ (exact-list)
+               (define literal-list (map Exact-v exact-list))
+               (define seen-any?
+                 (for/or ([literal (in-list literal-list)])
+                   (hash-ref seen literal #f)))
+
+               ;; We want to unconditionally iterate the whole list to mark
+               ;; all terms as seen.
+               ;;
+               ;; Consider:
+               ;;
+               ;; [1 ...]
+               ;; [(or 1 2) ...]
+               ;; [(or 2 3) ...]
+               ;;
+               ;; In this case,
+               ;; - 1 can be case-dispatched.
+               ;; - (or 1 2) can't because of duplicate 1, so it must be matched
+               ;; - (or 2 3) also can't because we just moved (or 1 2)
+               ;;   to the else clause. Since matching on 2 should reach
+               ;;   (or 1 2), not (or 2 3), (or 2 3) must be moved to the else
+               ;;   clause too.
+
+               (for ([literal (in-list literal-list)])
```

Comment:
Can we combine these two for loops into one?

Reply:
Not straightforwardly. Consider:

```
(match ...
  [(or 1 2 2) ...]
  [3 ...])
```

If we query and mark in the same loop, the the optimization would be inhibited because of duplicate `2`: when we see the second `2`, `2` is already marked.

Separating into two loops fixes the problem. The querying loop would say the `(or 1 2 2)` is OK. And only then, the marking loop would mark `1` and `2`.

So it adds more optimization opportunities.

Reply:
Makes sense. Can you make sure there's a test that ensures this case is optimized?

Reply:
It's already there. See test-case "Duplicate in a clause".